tactics/tactics.ml

```diff
@@ -3074,17 +3074,17 @@ let expand_hyp id = Tacticals.New.tclTRY (unfold_body id) <*> clear [id]
 
  *)
 
-let warn_unused_intro_pattern =
+let warn_unused_intro_pattern env =
   CWarnings.create ~name:"unused-intro-pattern" ~category:"tactics"
          (fun names ->
           strbrk"Unused introduction " ++ str (String.plural (List.length names) "pattern")
           ++ str": " ++ prlist_with_sep spc 
          (Miscprint.pr_intro_pattern 
-	 (fun c -> Printer.pr_econstr (snd (c (Global.env()) Evd.empty)))) names)
+	 (fun c -> Printer.pr_econstr (snd (c env Evd.empty)))) names)
 
-let check_unused_names names =
+let check_unused_names names env =
   if not (List.is_empty names) then
-    warn_unused_intro_pattern names
+    warn_unused_intro_pattern env names
```

Comment:
Downstream, the `sigma` should be passed to `pr_econstr` in replacement of `Evd.empty`. (The warning was actually twice wrong: in using `Global.env()` instead of the current goal environnement, and in not using the current `sigma` of evars). Upstream, the sigma should be obtained by `Proofview.Goal.sigma`, and this apparently requires to insert a `Proofview.Goal.enter` in the `[]` branch of `induct_discharge`.

By the way, for the same reason as for the `sigma`, the `env` in this `[]` branch should not be obtained by `tclENV` but by `Proofview.Goal.env` since it should be the env of the goal (containing the goal variables), and not the env of the proof state, which is only a local `Global.env()`.

I'm unsure of what you know and what you don't know, so, if you need more information about `sigma`, don't hesitate to ask.